src/lib.rs

```diff
@@ -1,47 +1,78 @@
 #![no_std]
+use bit_field::BitField;
+use embedded_hal::blocking::i2c::{Write, WriteRead};
+use num_enum::IntoPrimitive;
+use paste::paste;
 
-//! Manages an MCP23017, a 16-Bit I2C I/O Expander with Serial Interface module.
-//!
-//! This operates the chip in `IOCON.BANK=0` mode, i.e. the registers are mapped sequentially.
-//! This driver does not set `IOCON.BANK`, but the factory default is `0` and this driver does
-//! not change that value.
-//!
-//! See [the datasheet](http://ww1.microchip.com/downloads/en/DeviceDoc/20001952C.pdf) for more
-//! information on the device.
+/// Pin modes.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Direction {
+    /// Represents input mode.
+    Input,
+    /// Represents output mode.
+    Output,
+}
 
-#![deny(
-    missing_docs,
-    missing_debug_implementations,
-    missing_copy_implementations,
-    trivial_casts,
-    trivial_numeric_casts,
-    unstable_features,
-    unused_import_braces,
-    unused_qualifications,
-    warnings
-)]
-#![allow(dead_code, non_camel_case_types)]
-#![allow(clippy::uninit_assumed_init, clippy::upper_case_acronyms)]
+/// Pin levels.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Level {
```

Comment:
If this is onle Low/High, why not just use a `bool`?

Reply:
Couple of reasons:

* The mapping between `Low`/`High` and `bool` is not clear. It's not clear that `true` is `High`. E.g.
what is `set_gpio(OscillatorEnableN, false)` supposed to do? Will the oscillator be enabled or not?
* Since rust doesn't have keyword arguments, it's often not clear what a lone `false` will be the argument for. Imagine e.g. this type being wrapped and passed through a bit more API layers with more complex function signatures.

Reply:
To me, it's exceptionally clear that `set_gpio(true)` means to assert (logic level 1, VCC) on the GPIO signal. I may be biased because of my C-based background, but this is ultimately just a personal preference thing. Either approach seems valid to me.

Reply:
To assert a signal doesn't mean it's electrically high. See the example I gave, or any open collector usage, e.g. I2C, overtemperature signals, reset lines.

Reply:
And this style is also consistent with the PACs where enabled/disabled/asserted/deasserted/active/inactive/high/low are enums and not bools.